torch/_decomp/decompositions.py

```diff
@@ -3142,13 +3143,136 @@ def gru_impl(
+
+
+def aa_linear_filter(x):
+    x = torch.abs(x)
+    return 1.0 - torch.clamp(x, max=1.0)
+
+
+def _compute_indices_weights_aa(
+    out_size, in_size, scale, interp_size, align_corners, device
+):
+    scale = _compute_scale(in_size, out_size, align_corners, scale=scale)
+
+    support = (interp_size * 0.5) * scale if scale >= 1.0 else interp_size * 0.5
+    max_interp_size = int(math.ceil(support)) * 2 + 1
+
+    i = torch.arange(out_size, dtype=torch.long, device=device)
+
+    center = scale * (i + 0.5)
+    invscale = 1.0 / scale if scale >= 1.0 else 1.0
+
+    # compute source indices as [xmin, xmin+1, ..., xmin+xsize-1]
+    xmin = torch.clamp((center - support + 0.5).to(torch.long), min=0)
+    xsize = torch.clamp((center + support + 0.5).to(torch.long), max=in_size) - xmin
+    xsize = torch.clamp(xsize, max=max_interp_size)
+
+    # compute weights
+    j = torch.arange(max_interp_size, dtype=torch.long, device=device).view(-1, 1)
+    # TODO: use a generic function aa_filter defined for bilinear and bicubic
+    weights = aa_linear_filter((j + xmin - center + 0.5) * invscale)
+    weights = torch.where(j < xsize, weights, 0.0)
+    total_weights = weights.sum(dim=0)
+    weights = weights / total_weights
+
+    return xmin, weights
+
+
+def _separable_upsample_bilinear2d_aa_single_dim(
+    in_tensor, out_size, interp_dim, align_corners, scale=None
+):
+    # Assume that in_tensor dtype is float32
+    assert interp_dim % 4 in (2, 3)
+
+    n, c, in_h, in_w = in_tensor.shape
+    interp_size = 2  # bilinear
+    in_size = in_tensor.shape[interp_dim]
+
+    memory_format = utils.suggest_memory_format(in_tensor)
+
+    n_idx = torch.arange(n, device=in_tensor.device).view(n, 1, 1, 1)
+    c_idx = torch.arange(c, device=in_tensor.device).view(1, c, 1, 1)
+
+    src_idx_min, weights = _compute_indices_weights_aa(
+        out_size,
+        in_size,
+        scale,
+        interp_size,
+        align_corners,
+        device=in_tensor.device,
     )
+    max_interp_size = len(weights)
+
+    if interp_dim % 4 == 3:
+        # horizontal pass
+        in_y = torch.arange(in_h, device=in_tensor.device).view((1, 1, in_h, 1))
+        src_idx_min = src_idx_min.view(1, 1, 1, out_size)
+
+        in_tensor_list = [
+            in_tensor[
+                n_idx, c_idx, in_y, torch.clamp(src_idx_min + k, max=in_w - 1)
+            ].contiguous(memory_format=memory_format)
+            for k in range(max_interp_size)
+        ]
+    else:
+        # vertical pass
+        assert interp_dim % 4 == 2
+        in_x = torch.arange(in_w, device=in_tensor.device).view((1, 1, 1, in_w))
+        src_idx_min = src_idx_min.view(1, 1, out_size, 1)
+
+        in_tensor_list = [
+            in_tensor[
+                n_idx, c_idx, torch.clamp(src_idx_min + k, max=in_h - 1), in_x
+            ].contiguous(memory_format=memory_format)
+            for k in range(max_interp_size)
+        ]
```

Comment:
This is *surely* causing a massive perf hit. The way to do this is via some broadcasting over the indices